snuba/clusters/cluster.py

```diff
@@ -0,0 +1,104 @@
+from abc import ABC, abstractmethod
+from typing import Set
+
+from snuba import settings
+from snuba.clickhouse.native import ClickhousePool, NativeDriverReader
+from snuba.clickhouse.query import ClickhouseQuery
+from snuba.clusters.storage_sets import StorageSetKey
+from snuba.reader import Reader
+
+
+class Cluster(ABC):
+    """
+    A cluster is responsible for managing a collection of database nodes.
+
+    Clusters are configurable, and will be instantiated based on user defined settings.
+
+    Each storage must be mapped to a cluster via a storage set, which defines
+    the storages that must be located on the same cluster.
+
+    In future, clusters will also be responsible for co-ordinating commands that
+    need to be run on multiple hosts that are colocated within the same cluster.
+    The cluster will expose methods for:
+        - bootstrap
+        - migrate
+        - cleanup
+        - optimize
+    """
+
+    def __init__(self, storage_sets: Set[str]):
+        self.__storage_sets = storage_sets
+
+    def get_storage_sets(self) -> Set[StorageSetKey]:
+        return {StorageSetKey(storage_set) for storage_set in self.__storage_sets}
+
+    @abstractmethod
+    def get_reader(self) -> Reader[ClickhouseQuery]:
```

Comment:
well, indeed, making the Reader clickhouse specific means this does not work anymore because the Cluster interface becomes clickhouse specific.
You can make this class generic itself and based on the abstract generic reader class and materialize the parameter type in the clickhouse implementation.

```
class Cluster(ABC, Generic[TQuery]):
     def get_reader(self) -> Reader[TQuery]

class ClickhouseCluster(Cluster[ClickhouseQuery]):
     def get_reader(self) -> Reader[ClickhouseQuery]

```

I think the callsites should still type check because the everywhere this is used, we already reference the ClickhouseCluster explicitly

Reply:
Ah, thank you.